### PR TITLE
S3218: Add support for record structs

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 1]
```diff
@@ -204,14 +188,18 @@ private static void CheckProperty(SymbolAnalysisContext context, INamedTypeSymbo
...
-            SymbolAnalysisContext context, Func<T, Location> locationSelector, string memberType) where T : ISymbol
+        private static void CheckEventOrMethod<T>(T eventOrMethod,
+                                                  INamedTypeSymbol containerClassSymbol,
+                                                  SymbolAnalysisContext context,
+                                                  Func<T, Location> locationSelector,
+                                                  string memberType) where T : ISymbol
         {
-            var shadowsMethod = GetSelfAndOuterClasses(containerClassSymbol)
-                .SelectMany(c=> c.GetMembers(eventOrMethod.Name))
+            var selfAndOutterClasses = GetSelfAndOuterClasses(containerClassSymbol);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:197]: I introduced the local `selfAndOutterClasses` to avoid the creation of the same list twice. To make clear, that `selfAndOutterClasses` isn't lazy, I also changed the type from `IEnumerable<T>` to `IReadOnlyList<T>`.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 2]
```diff
@@ -47,42 +47,34 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                        !containerClassSymbol.IsClassOrStruct())
                     {
                         return;
                     }
 
                     var members = innerClassSymbol.GetMembers().Where(member => !member.IsImplicitlyDeclared);
                     foreach (var member in members)
                     {
-                        if (member is IPropertySymbol property)
+                        switch (member)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:59]: Row 38: Sould be renamed to `Rule` and moved to a single line

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 3]
```diff
@@ -47,42 +47,34 @@ protected override void Initialize(SonarAnalysisContext context)
...
-                        if (member is IMethodSymbol method)
-                        {
-                            CheckMethod(c, containerClassSymbol, method);
-                            continue;
-                        }
-
-                        if (member is INamedTypeSymbol namedType)
-                        {
-                            CheckNamedType(c, containerClassSymbol, namedType);
+                            case IPropertySymbol property:
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:61]: Line 35: Const can be `private`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 4]
```diff
@@ -47,42 +47,34 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                                CheckProperty(c, containerClassSymbol, property);
+                                break;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:63]: I would actually remove the `members` variable like this, with renaming `x`:
```
                    foreach (var member in innerClassSymbol.GetMembers().Where(x => !x.IsImplicitlyDeclared))
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 5]
```diff
@@ -47,42 +47,34 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                            case IEventSymbol @event:
+                                CheckEvent(c, containerClassSymbol, @event);
+                                break;
+                            case IMethodSymbol method:
+                                CheckMethod(c, containerClassSymbol, method);
+                                break;
+                            case INamedTypeSymbol namedType:
+                                CheckNamedType(c, containerClassSymbol, namedType);
+                                break;
+                            default:
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:76]: I'd be fine with not having the `default` here too. And ignoring the rule that will ask to have it. It's just a noise.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
We can go further with the cleanup

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 6]
```diff
@@ -119,22 +111,17 @@ protected override void Initialize(SonarAnalysisContext context)
             }
         }
 
-        private static void CheckMethod(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol, IMethodSymbol method)
-        {
-            CheckEventOrMethod(method, containerClassSymbol, context,
-                e =>
+        private static void CheckMethod(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol, IMethodSymbol method) =>
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:114]: If changed it into a switch statement to be inline with the other "switch statement in a foreach loop" in this file.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 7]
```diff
@@ -151,39 +138,36 @@ private static void CheckEvent(SymbolAnalysisContext context, INamedTypeSymbol c
...
+        private static void CheckProperty(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol, IPropertySymbol property) =>
             CheckFieldOrProperty(property, containerClassSymbol, context, p =>
             {
                 var reference = p.DeclaringSyntaxReferences.FirstOrDefault();
                 if (reference == null)
                 {
                     return null;
                 }
-                if (!(reference.GetSyntax() is PropertyDeclarationSyntax syntax))
+                if (reference.GetSyntax() is not PropertyDeclarationSyntax syntax)
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:159]: The `Func<ISymbol, Location> locationSelector` parameter is defined quite a lot here for all kinds of symbols. We could implement it as an extension method so it could be reused in other places too:
```CS
public static SyntaxToken? GetIdentifierToken(this ISymbol symbol) =>
  symbol.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() switch
  {
    PropertyDeclarationSyntax p => p.Identifier,
    VariableDeclaratorSyntax v => v.Identifier,
    // ..
   // Could be extended to other declarations too
   MethodDeclarationSyntax m => m.Identifier,
    _ => null,
  };
```
By doing so, the `Func<ISymbol, Location> locationSelector` parameter is not needed anymore. It could also be reused in `CheckNamedType` to get rid of the switch statement there.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
There should be a better way of dealing with the coverage. Similar functionality is available in `CSharpFacade.Instance.Syntax.NodeIdentifier`

**Comment by martin-strecker-sonarsource** [comment]
CI fails. Please have a look. Besides that, the PR is ready for review.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs [hunk 8]
```diff
@@ -89,5 +91,28 @@ private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, Syn
 
             return root.DescendantNodes();
         }
+
+        /// <summary>
+        /// Returns the Identifier token of the first declaring <see cref="SyntaxReference"/>.
+        /// For e.g. an <see cref="IPropertySymbol"/> this is <see cref="PropertyDeclarationSyntax.Identifier"/>.
+        /// </summary>
+        /// <returns>The idenifier token or <see langword="null"/> if there is no DeclaringSyntaxReference or the <see cref="SyntaxNode"/> does not have an Identifier.</returns>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs:99]: We generally don't write docs like this. You can remove all of them. The name and signature should express that.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs [hunk 9]
```diff
@@ -89,5 +91,28 @@ private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, Syn
...
+        public static SyntaxToken? GetFirstIdentifier(this ISymbol symbol) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs:100]: To replace the docs:
```suggestion
        public static SyntaxToken? FirstDeclaringReferenceIdentifier(this ISymbol symbol) =>
```
I generally don't like the `Get` prefix, because everything non-void is by-definition `get` or `something and get`. And it postpones understanding the purpose 3 characters later.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs [hunk 10]
```diff
@@ -89,5 +91,28 @@ private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, Syn
...
+            symbol.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax().NodeIdentifier();
+
+        /// <summary>
+        /// Returns the Identifier tokens of all declaring <see cref="SyntaxReference"/>s.
+        /// For e.g. an <see cref="INamedTypeSymbol"/> this can be a list of partial <see cref="ClassDeclarationSyntax.Identifier"/>.
+        /// </summary>
+        public static ImmutableArray<SyntaxToken> GetIdentifiers(this ISymbol symbol) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs:107]: Having this, the other one (`FirstDeclaringReferenceIdentifier`) should be just something like `symbol.DeclaringReferenceIdentifiers().FirstOrDefault()`, shouldn't it?
```suggestion
        public static ImmutableArray<SyntaxToken> DeclaringReferenceIdentifiers(this ISymbol symbol) =>
```

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs [hunk 11]
```diff
@@ -89,5 +91,28 @@ private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, Syn
...
+            (from r in symbol.DeclaringSyntaxReferences
+             select r.GetSyntax().NodeIdentifier() into i
+             where i != null
+             select (SyntaxToken)i)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs:111]: Please use the lambda notation

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs [hunk 12]
```diff
@@ -89,5 +91,28 @@ private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, Syn
...
+            .ToImmutableArray();
+
+        /// <inheritdoc cref="Helpers.Facade.SyntaxFacade{TSyntaxKind}.NodeIdentifier"/>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs:114]: Should not be needed

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs [hunk 13]
```diff
@@ -89,5 +91,28 @@ private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, Syn
...
+        private static SyntaxToken? NodeIdentifier(this SyntaxNode node)
+            => CSharpFacade.Instance.Syntax.NodeIdentifier(node);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs:116]: If you need this extension, then the responsibility should be the other way around. The `Facade` is a dispatcher to language-specific implementation.

Please move the implementation from the `Facade` here. And call this extension from the facade. And do the asymmetric step for VB to keep the codebase consistent.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 14]
```diff
@@ -102,135 +88,60 @@ protected override void Initialize(SonarAnalysisContext context)
                 return;
             }
 
-            foreach (var reference in namedType.DeclaringSyntaxReferences)
+            var namedTypeName = namedType.TypeKind switch { TypeKind.Class => "class", TypeKind.Delegate => "delegate" };
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:91]: From semantic of the algorithm, there's a question `else what?` The extraction was a good catch, but should follow the original idea:
```suggestion
            var namedTypeName = namedType.TypeKind == TypeKind.Delegate ? "delegate" : "class";
```

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:91]: `namedTypeName` sounds too wordy. Can it be something like `kindName` instead?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 15]
```diff
@@ -102,135 +88,60 @@ protected override void Initialize(SonarAnalysisContext context)
...
                 .OfType<IPropertySymbol>()
                 .Any(prop => prop.IsStatic);
-            var shadowsField = GetSelfAndOuterClasses(containerClassSymbol)
+            var shadowsField = selfAndOutterClasses
                 .SelectMany(c => c.GetMembers(propertyOrField.Name))
                 .OfType<IFieldSymbol>()
                 .Any(field => field.IsStatic || field.IsConst);
 
-            if (shadowsProperty || shadowsField)
+            if ((shadowsProperty || shadowsField) && propertyOrField.GetFirstIdentifier()?.GetLocation() is { } location)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:113]: Let's exract only the `identifier` here, as the `GetLocation()` will never return `null` for a known token. Current version implies that it can happen.
```suggestion
            if ((shadowsProperty || shadowsField) && propertyOrField.GetFirstIdentifier() is { } identifier)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 16]
```diff
@@ -102,135 +88,60 @@ protected override void Initialize(SonarAnalysisContext context)
...
-
-                if (syntax is ClassDeclarationSyntax classSyntax)
-                {
-                    context.ReportDiagnosticIfNonGenerated(Diagnostic.Create(rule, classSyntax.Identifier.GetLocation(), "class"));
-                }
+                context.ReportDiagnosticIfNonGenerated(Diagnostic.Create(Rule, identifier.GetLocation(), namedTypeName));
             }
         }
 
-        private static void CheckMethod(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol, IMethodSymbol method)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:122]: This extension was a great catch!

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 17]
```diff
@@ -102,135 +88,60 @@ protected override void Initialize(SonarAnalysisContext context)
...
-                }
-                return syntax.Identifier.GetLocation();
-            }, "property");
-        }
-
-        private static void CheckFieldOrProperty<T>(T propertyOrField, INamedTypeSymbol containerClassSymbol,
-            SymbolAnalysisContext context, Func<T, Location> locationSelector, string memberType) where T : ISymbol
-        {
-            var shadowsProperty = GetSelfAndOuterClasses(containerClassSymbol)
+            var selfAndOutterClasses = GetSelfAndOuterClasses(containerClassSymbol);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:103]: From a performance PoV: This is called for each member. Everyone does `GetSelfAndOuterClasses`. 
From a duplication PoV: Everyone also does `.SelectMany(c => c.GetMembers(xxx.Name))` for each member.

Let's move the classes extraction before the `foreach` in `Initialize` (line 53). And the `GetMembers` thing before the switch (line 55)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 18]
```diff
@@ -102,135 +88,60 @@ protected override void Initialize(SonarAnalysisContext context)
...
                 .OfType<IMethodSymbol>()
                 .Any(method => method.IsStatic);
-            var shadowsEvent = GetSelfAndOuterClasses(containerClassSymbol)
+            var shadowsEvent = selfAndOutterClasses
                 .SelectMany(c => c.GetMembers(eventOrMethod.Name))
                 .OfType<IEventSymbol>()
                 .Any(@event => @event.IsStatic);
 
-            if (shadowsMethod || shadowsEvent)
+            if ((shadowsMethod || shadowsEvent) && eventOrMethod.GetFirstIdentifier()?.GetLocation() is { } location)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:134]: Using `x` in these lambdas can fix the `@event` collision pain

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs (hunk 17 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:103]: And also `IsStatic` seems to be wanted by everyone, so it can be extracted. It will need extra glitch for IFieldSymbol and IsConst, something like `(...IsStatic || (... is IFieldSymbol {IsConst: true})`

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Round 2. The rule is now greatly simplified, maybe some more pattern will emerge once we do it. I have a feeling that `CheckFieldAndProperty` and `CheckEventOrMethod` are extremely similar to each other and might be mergable with some generic magic.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs (hunk 13 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs:116]: I changed the VB implementation and merged some extensions methods in VB. We should do the same in C# if the implementation is fine in VB.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs (hunk 17 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:103]: I merged all into CheckFieldPropertyEventOrMethod (CheckNamedType is still on its own).

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 19]
```diff
@@ -25,212 +25,98 @@
...
+                            case IPropertySymbol property:
+                                CheckFieldPropertyEventOrMethod(property, containerClassSymbol, c, "property");
+                                break;
+                            case IFieldSymbol field:
+                                CheckFieldPropertyEventOrMethod(field, containerClassSymbol, c, "field");
+                                break;
+                            case IEventSymbol @event:
+                                CheckFieldPropertyEventOrMethod(@event, containerClassSymbol, c, "event");
+                                break;
+                            case IMethodSymbol { MethodKind: MethodKind.DeclareMethod or MethodKind.Ordinary } method:
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:67]: There are a lot of IMethodSymbols, that can not collide, because they are unnamed (and they don't have a location in code).

**Review by martin-strecker-sonarsource — COMMENTED** [review]
Addressed your comments. VB and C# Facts and Extensions are a bit out of sync now. We should discuss offline, what the best way forward is here.

**Comment by martin-strecker-sonarsource** [comment]
ITs are reporting a new positive for the rule in Akka.net. Can we look into that together?

https://dev.azure.com/sonarsource/DotNetTeam%20Project/_build/results?buildId=46751&view=logs&j=c7ac5d00-ca19-592b-a6ee-71358b142e5c&t=d7fd8ed3-4895-531e-2ba8-7ed195398253

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 20]
```diff
@@ -25,212 +25,98 @@
...
-        private static void CheckNamedType(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol,
-            INamedTypeSymbol namedType)
+        private static void CheckNamedType(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol, INamedTypeSymbol namedType)
         {
             var shadowsClassOrDelegate = GetSelfAndOuterClasses(containerClassSymbol)
-                .SelectMany(c => c.GetMembers(namedType.Name))
-                .OfType<INamedTypeSymbol>()
-                .Any(nt => nt.Is(TypeKind.Class) || nt.Is(TypeKind.Delegate));
+                .SelectMany(x => x.GetMembers(namedType.Name))
+                .Any(x => x is INamedTypeSymbol symbol && symbol.TypeKind is TypeKind.Class or TypeKind.Delegate);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:83]: TODO: Check test cases for struct

**Comment by martin-strecker-sonarsource** [comment]
@pavel-mikula-sonarsource  This comment still applies: https://github.com/SonarSource/sonar-dotnet/pull/5577#pullrequestreview-951598308

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs (hunk 15 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs]: No, we should discover that once it is in the language as part of the specification sprint for new language features.

File: analyzers/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs [hunk 21]
```diff
@@ -22,12 +22,15 @@
...
 using System.Collections.Immutable;
 using System.Linq;
+using System.Reflection;
 using Microsoft.CodeAnalysis;
 
 namespace SonarAnalyzer.Helpers
 {
     internal static class SymbolHelper
     {
+        private static readonly PropertyInfo IsRecordPropertyInfo = typeof(ITypeSymbol).GetProperty("IsRecord");
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs:32]: The `PropertyInfo` part is visible from the type. It's more relevant what type it relates to.
```suggestion
        private static readonly PropertyInfo ITypeSymbolIsRecord = typeof(ITypeSymbol).GetProperty("IsRecord");
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Round 3, minor improvements

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 22]
```diff
@@ -25,212 +25,92 @@
...
+                                break;
                         }
                     }
                 },
                 SymbolKind.NamedType);
         }
 
-        private static void CheckNamedType(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol,
-            INamedTypeSymbol namedType)
+        private static void CheckNamedType(SymbolAnalysisContext context, INamedTypeSymbol containerClassSymbol, INamedTypeSymbol namedType)
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:73]: > It has fixed result that is recomputed for each member

🤦‍♂️ I completely missed that. Great catch.

**Comment by martin-strecker-sonarsource** [comment]
And please let's not forget about: https://github.com/SonarSource/sonar-dotnet/pull/5577#issuecomment-1109715489

File: analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S3218.json [hunk 23]
```diff
@@ -0,0 +1,17 @@
+{
+"issues":  [
+{
+"id":  "S3218",
+"message":  "Rename this field to not shadow the outer class' member with the same name.",
+"location":  {
+"uri":  "sources\akka.net\src\core\Akka.Streams\Implementation\ActorMaterializerImpl.cs",
```
Comment (martin-strecker-sonarsource) [inline analyzers/its/expected/akka.net/Akka.Streams--netstandard2.0-S3218.json:7]: True positive. There is a static `Props` method in the class that contains the class that contains the marked property.

```cs
    public class StreamSupervisor : ActorBase
    {
        public sealed class Materialize : INoSerializationVerificationNeeded, IDeadLetterSuppression
        {
            public readonly Props Props; // Line 537
        }
        public static Props Props(ActorMaterializerSettings settings, AtomicBoolean haveShutdown) // Line 626
            => Actor.Props.Create(() => new StreamSupervisor(settings, haveShutdown)).WithDeploy(Deploy.Local);
    }
```

**Comment by martin-strecker-sonarsource** [comment]
Quality gate was complaining about a missing default in a switch. I marked it as don' fix here https://sonarcloud.io/project/issues?pullRequest=5577&issues=AYCET4AsFbD8hslytJ7p&open=AYCET4AsFbD8hslytJ7p&id=sonaranalyzer-dotnet

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 24]
```diff
@@ -25,212 +25,92 @@
...
             var shadowsClassOrDelegate = GetSelfAndOuterClasses(containerClassSymbol)
-                .SelectMany(c => c.GetMembers(namedType.Name))
-                .OfType<INamedTypeSymbol>()
-                .Any(nt => nt.Is(TypeKind.Class) || nt.Is(TypeKind.Delegate));
+                .SelectMany(x => x.GetMembers(namedType.Name))
+                .Any(x => x is INamedTypeSymbol symbol && symbol.TypeKind
+                    is TypeKind.Class
+                    or TypeKind.Struct
+                    or TypeKind.Delegate
+                    or TypeKind.Enum);
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs]: The first syntax sugar about pattern matching was not addressed.

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs [hunk 25]
```diff
@@ -89,5 +91,33 @@ private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, Syn
...
+            symbol.DeclaringReferenceIdentifiers().FirstOrDefault();
+
+        public static ImmutableArray<SyntaxToken> DeclaringReferenceIdentifiers(this ISymbol symbol) =>
+            symbol.DeclaringSyntaxReferences
+               .Select(x => x.GetSyntax().NodeIdentifier())
+               .WhereNotNull()
+               .ToImmutableArray();
+
+        public static SyntaxToken? NodeIdentifier(this SyntaxNode node)
+            => node.RemoveParentheses() switch
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/ISymbolExtensions.cs:105]: Formatting nitpick `=>`
```suggestion
        public static SyntaxToken? NodeIdentifier(this SyntaxNode node) => 
            node.RemoveParentheses() switch
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 26]
```diff
@@ -25,217 +25,91 @@
...
-            INamedTypeSymbol namedType)
+        private static void CheckNamedType(SymbolAnalysisContext context, IReadOnlyList<ISymbol> outterMembersOfSameName, INamedTypeSymbol namedType)
         {
-            var shadowsClassOrDelegate = GetSelfAndOuterClasses(containerClassSymbol)
-                .SelectMany(c => c.GetMembers(namedType.Name))
-                .OfType<INamedTypeSymbol>()
-                .Any(nt => nt.Is(TypeKind.Class) || nt.Is(TypeKind.Delegate));
-
-            if (!shadowsClassOrDelegate)
+            if (!outterMembersOfSameName.Any(x => x is INamedTypeSymbol symbol && symbol.TypeKind is TypeKind.Class or TypeKind.Struct or TypeKind.Delegate or TypeKind.Enum))
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:82]: I'd revert the condition to be
```
if(...)
{
    foreach(...
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs [hunk 27]
```diff
@@ -25,217 +25,91 @@
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Extensions;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MemberShadowsOuterStaticMember.cs:28]: Line 21: `using System;` can be removed

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM, there are some nitpicks left - and one unresolved conversation above about pattern matching (hidden somewhere in GitHub UI)
